### DISCUSSION TITLE
Oqtane 3.1

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#2061] [General]
Oqtane 3.1 is focused on a variety of security enhancements which have been on the roadmap for quite some time. I merged a large PR today which contains support for OpenID Connect authentication - however it is still a work in progress... so please be aware that the Dev branch may have some instability until this feature is complete (hopefully by the end of this week).

**Comment by curlyfro** [comment]
YAY!!!  testing now.

**Reply by hishamco** [reply]
Is there other than OpenID should be shipped in 3.1?

**Reply by sbwalker** [reply]
@hishamco the goal is to get OIDC working first so that any OIDC provider can be used for authentication. This is a key aspect of the BFF architecture I outlined in this thread https://github.com/oqtane/oqtane.framework/discussions/2012

@curlyfro it may be better to wait a few days for testing as it is still quite rough around the edges - which was why I posted the PSA about the Dev branch.

**Comment by sbwalker** [comment]
The Dev branch should be stable now as the OIDC enhancements are now complete

**Comment by sbwalker** [comment]
The specific security enhancements which were in scope for the 3.1 release are outlined below:

- user account lockout support
- two factor authentication support
- password complexity configuration
- per-site configuration of authentication options
- external login support via OAuth2 / OpenID Connect
- federated / SSO authentication support via OpenID Connect
- auth token support for allowing external clients to call APIs
- auth token support for calling remote/downstream APIs

The functional requirements for these enhancements included:

- must work on Blazor Server and Blazor WebAssembly
- must support multi-tenancy (ie. site-based configuration)
- must support all current database providers
- must allow for global configuration in appsettings
- must allow for management via the administrative user interface
- must not adversely impact application performance
- must be backward compatible and include no breaking changes

Site Options

The challenge which took the most effort to solve was related to multi-tenancy support. Multi-tenancy is not a first class citizen in .NET Core and as a result it usually requires a decent amount of creativity in order to provide support. Luckily other clever developers have encountered this problem previously and have shared their knowledge and approach with the community. 

In particular, the configuration aspects of .NET Core rely on the options pattern. By default the options pattern was not designed for multi-tenancy however there is a need in Oqtane to be able to override the default application options with site-specific options. Luckily the options pattern does provide some extensibility so it is possible to alter the default behavior. The end result is a solution in Oqtane that allows options to be stored per site in the database which will override the default application options at run-time. And the nice thing is that the per-site options will work with any options class that supports IOptions<TOptions> which should simplify many future integrations.

As part of the site options implementation some new concepts were added which should simplify multi-tenancy development. The most important concept in multi-tenant development is determining the current tenant. In prior versions of Oqtane the most common method was to inject the TenantManager and call GetAlias (which would provide access to the TenantId, SiteId, etc...). In Oqtane 3.1 the tenant information is now added automatically to the HttpContext and an extension method has been added so that it is as simple as calling HttpContext.GetAlias within an API Controller. Or if you are in a class where the HttpContext is not available you can now inject an IAliasAccessor service and reference the Alias property.

.NET Core Identity

Traditionally the primary authentication method available in Oqtane was based on logging in via a local user account. This functionality relies on the standard .NET Core Identity provider however Oqtane used a customized implementation to enable the user interface to be written using native Blazor components. However the customized implementation did not provide full support for all of the .NET Core Identity capabilities, so as a result the 3.1 release has added a variety of additional features in this area. These were added to the Settings tab of the User Management page.

![image](https://user-images.githubusercontent.com/4840590/160867961-cc3cf8c3-88ee-4886-b177-2c424ee3e6ff.png)

Password Complexity

Ensuring your users utilize sufficiently complex passwords is vital to the integrity of your site. The default password criteria has been significantly hardened in 3.1 and there is now the ability to customize the criteria per site as well. Increased password complexity makes it more difficult for users to enter their passwords correctly, so a feature has been added to Login to allow a user to view the password they entered if they are having difficulty.

![image](https://user-images.githubusercontent.com/4840590/160005644-2d2c8b64-7fce-4e58-9216-dc9acfc53640.png)

User Lockout

The user lockout feature is a way to improve application security by locking out a user that enters a password incorrectly several times. This technique can help  protect against brute force attacks where an attacker repeatedly tries to guess a password. This feature is now enabled by default in 3.1 and there is an ability to customize the criteria per site as well.

Two Factor

Two factor authentication (also called 2FA, multi-factor, or MFA) is a process in which a user is requested during a sign-in event for additional forms of identification. Requiring a second form of authentication significantly enhances security as the additional factor cannot be easily obtained or duplicated by an attacker. The 3.1 release now allows a user to specify if they wish to enable two factor authentication in their account. Please note that the Admin needs to enable this feature and the only supported delivery mechanism at this time is email notifications.

![image](https://user-images.githubusercontent.com/4840590/160005783-13404fa8-492b-4518-8485-4cee6283c4ac.png)

External Login

Externalizing the authentication for your site has a number of key benefits. It allows users to identify themselves across websites and apps without having to manage a local password for each site. In addition, utilizing a trusted identity provider is a good general security practice as the fewer people that have access to your login information, the better. It also provides the opportunity for single sign-on across multiple sites and can be used to federate security with third party services and APIs.

Oqtane 3.1 adds support for integrating with an OAuth 2.0 or OpenID Connect provider - both of which are international standards that are widely used for enterprise IT security. OAuth 2.0 is a framework that controls authorization to protected resources such as applications or websites. OpenID Connect is an industry standard for federated authentication that is built on top of the OAuth 2.0 protocol and is specifically focused on user authentication. The 3.1 release provides the ability to customize the External Login criteria per site.

![image](https://user-images.githubusercontent.com/4840590/160005749-9d3419b1-d3f3-4779-8b28-daf82f7b27e2.png)

It is important to note that External Login providers are for authentication purposes only. The External Login provider must provide a valid email address for the signed in user, which will will be mapped to a local user in the site. The local user will be assigned to roles to provide them with permissions on the site. Note that it is possible to have External Login enabled while still providing the ability for a user to login using a local account (this capability can be disabled as well).

![image](https://user-images.githubusercontent.com/4840590/160005823-567db100-f71f-4545-92d6-2886e950a87c.png)

Note that OpenID Connect is the preferred approach for remote authentication... however not all services are providing an OpenID Connect endpoint yet. However OpenID Connect is based on OAuth 2.0 - it just layers on an additional ID token to standardize areas that OAuth 2.0 leaves up to choice, such as scopes and endpoint discovery. So it is possible to use standard OAuth 2.0 for authentication - you just need to provide the 3 required Urls - auth url, token url, and user info url (and client id and secret). A good example of a service which does not yet support OpenID Connect is Github... however you can use OAuth 2.0 as an integration point:

Authorization Url: https://github.com/login/oauth/authorize
Token Url: https://github.com/login/oauth/access_token
User Info Url: https://api.github.com/user/emails (required as the https://api.github.com/user endpoint does not provide private email addresses)
Scopes: user:email
Domain Filter: !users.noreply.github.com  (this is required to ignore the Github public email address and capture the user's private email address)

**Comment by vzdesic** [comment]
@sbwalker Congratulations for the 3.1 release!

I am very excited to try this. I was waiting for OIDC implementation and I followed different discussions here in order to be up to date. It is great that it is here!

I would like to understand more details about OIDC implementation. Can we use any Identity Provider or there are some limitations and Oqtane covers only some of them? 

I am asking that because I remeber that there was a discussions about few of them; AAD B2C, Identity Server 4 and Openiddict at the end. Now, it seems that we are not limited to any of them, because proper flexible implementation of the OIDC protocol itself is now part of the Oqtane, right? 

It seems that the OIDC configuration included in this version is already capable to cover some range of scenarios including different Identity Providers. Can you say which IPs were used during development if any other than AAD B2C obviously? Can you say more about differencies between server-side and WebAssembly in this regard.

Thanks a lot!

**Reply by sbwalker** [reply]
@vzdesic  The 3.1 version has actually not been released yet... it is still being tested, but it is near feature complete. Here is an updated diagram of the overall security architecture:

![image](https://user-images.githubusercontent.com/4840590/160839848-287432f0-7b3f-46b7-bd20-5a703cf5c7fc.png)

You are correct in your observation that Oqtane did not include any dependencies on third party libraries such as IdentityServer, Duende, OpenIddict, or Microsoft's Web Identity. It is focusing on the open web standards and using the native capabilities in .NET Core. 

In terms of authentication it can connect to any OpenId Connect or OAuth 2.0 server. Examples that have been tested so far are Azure B2C, Auth0, Okta, IdentityServer, and Github. Basically, as long as the external auth server is able to return an email address for the user (which is one of the standard claims in OIDC) then integration is possible with Oqtane.

In terms of authorization, it is using the BFF architecture outlined in more detail in #2012 which relies on same site cookies. This works well for the main web client use case with APIs as part of the same domain. However there are other use cases to consider as well.

For the use case of an external client application which wants to connect to the Oqtane API, a Jwt authentication capability is now possible. A site administrator can configure their Jwt token details and generate a long-lived token which can be used to access the API.

![image](https://user-images.githubusercontent.com/4840590/160845820-2a815e65-8e46-49ab-8433-4562b5b12397.png)

Another use case is allowing the Oqtane web client to call remote API services which do not live in the same domain (ie. micro-services). A RemoteServiceBase class has been added which can be used by Razor components to make remote API calls (the RemoteServiceBase includes an HttpClient which has the auth token already configured). The remote API would simply need to use the standard AddJwtBearer() method in its Startup to validate the token and provide secure access to its API methods ie.

```
builder.Services.AddAuthentication(options => {
        options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
        options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
    })
    .AddJwtBearer(options => {
        options.RequireHttpsMetadata = false;
        options.SaveToken = true;
        options.TokenValidationParameters = new TokenValidationParameters()
        {
            ValidateIssuerSigningKey = true,
            ValidateIssuer = true,
            IssuerSigningKey = new SymmetricSecurityKey(System.Text.Encoding.UTF8.GetBytes("xxxxxxxxxxxxxxxx")),
            ValidIssuer = "http://localhost:44357",
            ValidateAudience = false,
            ValidateLifetime = true,
            ClockSkew = TimeSpan.FromDays(1),
        };
    });
```
and standard auth attributes in controllers:

```
    [HttpGet]
    [Authorize]
    //[Authorize(Roles = "Administrators")] - standard Role authorization also works using the roles assigned in Oqtane
    public IEnumerable<WeatherForecast> Get()
```

To answer your other question... these features work on both Blazor Server and Blazor WebAssembly. And the addition of token-based authentication ensures Oqtane is ready for future Blazor use cases such as running your components in .NET MAUI as Blazor Hybrid apps.

**Comment by Sharafudeen** [comment]
When we can get final release of Oqtane 3.1? We are eagerly waiting for these stable features! Also please post updated video tutorials. Thanks for your efforts!

**Comment by sbwalker** [comment]
Oqtane 3.1 is now released - https://www.oqtane.org/blog/!/41/oqtane-3-1-released

**Comment by larsontim12** [comment]
Congratulations! This is the release I have been waiting for.  You guys have really outdone yourselves with your commitment in getting this project to where it is now.

Thank you for all the hard work and effort!
Tim
